kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionManager.scala

```diff
@@ -198,10 +202,85 @@ class KyuubiSessionManager private (name: String) extends SessionManager(name) {
       ms.registerGauge(EXEC_POOL_ALIVE, getExecPoolSize, 0)
       ms.registerGauge(EXEC_POOL_ACTIVE, getActiveCount, 0)
     }
-    // TODO: support to recover batch sessions with session state store
+    recoverBatchSessions()
     super.start()
   }
 
+  @VisibleForTesting
+  def recoverBatchSessions(): Unit = {
+    val restHost = conf.get(FRONTEND_REST_BIND_HOST)
+      .getOrElse(Utils.findLocalInetAddress.getHostAddress)
+    val restPort = conf.get(FRONTEND_REST_BIND_PORT)
+    val kyuubiInstance = s"$restHost:$restPort"
+    val recoverPerBatch = conf.get(SERVER_STATE_STORE_RECOVERY_SESSIONS_PER_BATCH)
+
+    val batchSessions = ListBuffer[KyuubiBatchSessionImpl]()
+    Seq(OperationState.PENDING, OperationState.RUNNING).foreach { stateToRecover =>
+      var offset = 0
+      var lastRecoveryNum = Int.MaxValue
+
+      while (lastRecoveryNum >= recoverPerBatch) {
+        val metadataList = sessionStateStore.getBatchesRecoveryMetadata(
+          stateToRecover.toString,
+          kyuubiInstance,
+          offset,
+          recoverPerBatch)
+        metadataList.foreach { metadata =>
+          val batchRequest = new BatchRequest(
+            metadata.engineType,
+            metadata.resource,
+            metadata.className,
+            metadata.requestName,
+            metadata.requestConf.asJava,
+            metadata.requestArgs.asJava)
+
+          val batchSession = new KyuubiBatchSessionImpl(
+            BatchesResource.REST_BATCH_PROTOCOL,
+            metadata.username,
+            "",
+            metadata.ipAddress,
+            metadata.requestConf,
+            this,
+            conf.getUserDefaults(metadata.username),
+            batchRequest,
+            Some(metadata))
+
+          batchSessions += batchSession
+        }
+
+        lastRecoveryNum = metadataList.size
+        offset += lastRecoveryNum
+      }
+    }
+
+    batchSessions.foreach { batchSession =>
+      val user = batchSession.user
```

Comment:
just KyuubiSessionManager.openBatchSession ?

Reply:
we can not open the batch session during recovery, it will change the metadata state.

Reply:
we can recover RUNNING at first, then recover PENDING.

Reply:
after think twice, we can not reuse it, if the state is changed.

We might get duplicate bathes to recover.

Reply:
unless we revert the recovery order.

Recover 10 ~ 20 first, then recover 0 ~ 10.